ee/tabby-webserver/src/repositories/oauth.rs

```diff
@@ -0,0 +1,120 @@
+use std::sync::Arc;
+
+use anyhow::Result;
+use axum::{
+    extract::{Path, Query, State},
+    response::Redirect,
+    routing, Router,
+};
+use hyper::StatusCode;
+use juniper::ID;
+use serde::Deserialize;
+
+use crate::{
+    oauth::github::GithubOAuthResponse,
+    schema::{
+        github_repository_provider::GithubRepositoryProviderService, setting::SettingService,
+    },
+};
+
+#[derive(Clone)]
+pub struct OAuthState {
+    pub settings: Arc<dyn SettingService>,
+    pub github_repository_provider: Arc<dyn GithubRepositoryProviderService>,
+}
+
+pub fn routes(state: OAuthState) -> Router {
+    Router::new()
+        .route("/github/login/:id", routing::get(login))
+        .route("/github/callback", routing::get(callback))
+        .with_state(state)
+}
+
+fn github_redirect_url(client_id: &str, redirect_uri: &str, id: &ID) -> String {
+    format!("https://github.com/login/oauth/authorize?client_id={client_id}&response_type=code&scope=repo&redirect_uri={redirect_uri}/repositories/oauth/github/callback&state={id}")
+}
+
+#[derive(Deserialize)]
+struct CallbackParams {
+    state: ID,
+    code: String,
+}
+
+macro_rules! log_error {
```

Comment:
prefer not using a macro in this kind of case for readability. (and encourange user generate proper error code, instead of just internal_server_error).